### PR TITLE
Standalone Lite: Launchers

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Adds the launchers.

- The base class remains unchanged
- PL now uses the _Launcher base class from Lite
- Implementations are trimmed down to just what Lite needs. PL specific implementations remain where they are.
- Later on, PL can refactor their implementations to depend on these ones (e.g. through utilities) or generalizations without depending on a Trainer

All launchers are protected API. No back-compatibility steps required.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: src/lightning_lite/strategies/launchers/multiprocessing.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
+# Copyright The PyTorch Lightning team.
```
Comment (awaelchli) [inline src/lightning_lite/strategies/launchers/multiprocessing.py:1]: Reviewer note: The base class gets moved, but the implementations need to be copied due to dependencies on PL.

File: src/lightning_lite/strategies/launchers/multiprocessing.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+import torch.multiprocessing as mp
+from typing_extensions import Literal
+
+from lightning_lite.strategies.launchers.base import _Launcher
+from lightning_lite.utilities.apply_func import move_data_to_device
+from lightning_lite.utilities.imports import _TORCH_GREATER_EQUAL_1_11
+from lightning_lite.utilities.seed import _collect_rng_states, _set_rng_states
+
+
+class _MultiProcessingLauncher(_Launcher):
```
Comment (awaelchli) [inline src/lightning_lite/strategies/launchers/multiprocessing.py:30]: Note, the trainer reference and functionality to copy trainer specific attributes (results) to main process has been removed.
No other changes.

File: src/lightning_lite/strategies/launchers/subprocess_script.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        return False
+
+    # TODO(lite): Update type annotation once ClusterEnvironment has moved to Lite
+    def __init__(self, cluster_environment: "ClusterEnvironment", num_processes: int, num_nodes: int) -> None:
+        super().__init__()
+        self.cluster_environment = cluster_environment
+        self.num_processes = num_processes
+        self.num_nodes = num_nodes
+
+    def launch(self, function: Callable, *args: Any, **kwargs: Any) -> Any:
```
Comment (awaelchli) [inline src/lightning_lite/strategies/launchers/subprocess_script.py:78]: Trainer references removed. No other changes

File: src/lightning_lite/strategies/launchers/xla.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    """
+
+    def __init__(self, strategy: "Strategy") -> None:
+        super().__init__(strategy=strategy, start_method="fork")
+
+    @property
+    def is_interactive_compatible(self) -> bool:
+        return True
+
+    def launch(self, function: Callable, *args: Any, **kwargs: Any) -> Any:
```
Comment (awaelchli) [inline src/lightning_lite/strategies/launchers/xla.py:59]: Same removals as in super class: Removed _collect_rank_zero_results

File: tests/tests_lite/strategies/launchers/test_multiprocessing.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
+# Copyright The PyTorch Lightning team.
```
Comment (awaelchli) [inline tests/tests_lite/strategies/launchers/test_multiprocessing.py:1]: New unit tests added as I saw fit. In PL, the launchers are mainly tested through integration testing, which we can't yet do before all parts have moved to Lite.

**Review by carmocca — APPROVED** [review]

File: src/lightning_lite/strategies/launchers/subprocess_script.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+
+    Arguments:
+        cluster_environment: A cluster environment that provides access to world size, node rank, etc.
+        num_processes: The number of processes to launch in the current node.
+        num_nodes: The total number of nodes that participate in this process group.
+    """
+
+    @property
+    def is_interactive_compatible(self) -> bool:
+        return False
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/lightning_lite/strategies/launchers/subprocess_script.py:69]: 100% agree, that's an accident

**Review by Borda — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by justusschock — APPROVED** [review]
